### PR TITLE
feat: auto-select lean 3 mode

### PR DESCRIPTION
**Pull request by cipher1024**
Some people split their time between Lean 3 and Lean 4. This tweak uses `elan` to check which version of Lean is used for a particular file and select the appropriate emacs mode.

### CODE REVIEW

File: lean4-mode/lean4-mode.el [hunk 1]
```diff
@@ -208,9 +208,30 @@ Invokes `lean4-mode-hook'.
 ;;;###autoload
 (push '("\\.lean$" . lean4-mode) auto-mode-alist)
 
+(defun lean--version ()
+  (let ((version-line (car (process-lines "lean" "-v"))))
```
Comment (Kha) [inline lean4-mode/lean4-mode.el:212]: Please use `lean4-get-executable`. Also, have you tested this when elan first has to download Lean and prints its own output first?

File: lean4-mode/lean4-mode.el [hunk 2]
```diff
@@ -208,9 +208,30 @@ Invokes `lean4-mode-hook'.
...
+  (message "Lean %s" (mapconcat 'identity (lean--version) ".")))
+(defalias 'lean4-show-version 'lean-show-version)
+
+(defun lean-select-mode ()
+  (if lean4-autodetect-lean3
+      (let ((version (lean--version)))
+        (cond ((equal (car version) "4") (lean4-mode))
+              ((equal (car version) "3") (lean-mode))))
+    (lean4-mode)))
+(defalias 'lean4-select-mode 'lean4-select-mode)
```
Comment (Kha) [inline lean4-mode/lean4-mode.el:229]: Typo? Why are these necessary anyway?

File: lean4-mode/lean4-mode.el [hunk 3]
```diff
@@ -208,9 +208,30 @@ Invokes `lean4-mode-hook'.
 ;;;###autoload
 (push '("\\.lean$" . lean4-mode) auto-mode-alist)
```
Comment (Kha) [inline lean4-mode/lean4-mode.el:209]: Uh, don't you want to change this one too? How did you test this patch?

**Review by Kha — CHANGES_REQUESTED** [review]

File: lean4-mode/lean4-mode.el (hunk 1 above)
Comment (cipher1024) [reply lean4-mode/lean4-mode.el:212]: I hadn't thought of that. Is there a way of asking if elan if the required lean version needs to be downloaded?

File: lean4-mode/lean4-mode.el (hunk 2 above)
Comment (cipher1024) [reply lean4-mode/lean4-mode.el:229]: They are mostly cosmetic. I just think that this feature conceptually belongs to more name spaces. They're not very important and I can remove them.

File: lean4-mode/lean4-mode.el (hunk 3 above)
Comment (cipher1024) [reply lean4-mode/lean4-mode.el:209]: Good catch, thanks!

It is indeed poorly tested, I apologize. It is a version of something I have in my `init.el` file. In general, I don't have a systematic approach to testing my elisp code and any advice would be welcome.

File: lean4-mode/lean4-mode.el (hunk 1 above)
Comment (Kha) [reply lean4-mode/lean4-mode.el]: I think it's sufficient to just parse the last line instead

File: lean4-mode/lean4-mode.el (hunk 3 above)
Comment (Kha) [reply lean4-mode/lean4-mode.el]: I believe Soonho was the last person to know anything about unit-testing Elisp... :)
